// aws-jwt-verify/dist/esm/node-web-compat-node.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Node.js implementations for the node-web-compatibility layer
import { createPublicKey, createVerify, verify } from "crypto";
import { fetch } from "./https-node.js";
/**
 * Enum to map supported JWT signature algorithms with OpenSSL message digest algorithm names
 */
var JwtSignatureAlgorithmHashNames;
(function (JwtSignatureAlgorithmHashNames) {
    JwtSignatureAlgorithmHashNames["RS256"] = "RSA-SHA256";
    JwtSignatureAlgorithmHashNames["RS384"] = "RSA-SHA384";
    JwtSignatureAlgorithmHashNames["RS512"] = "RSA-SHA512";
    JwtSignatureAlgorithmHashNames["ES256"] = "RSA-SHA256";
    JwtSignatureAlgorithmHashNames["ES384"] = "RSA-SHA384";
    JwtSignatureAlgorithmHashNames["ES512"] = "RSA-SHA512";
})(JwtSignatureAlgorithmHashNames || (JwtSignatureAlgorithmHashNames = {}));
export const nodeWebCompat = {
    fetch,
    transformJwkToKeyObjectSync: (jwk) => createPublicKey({
        key: jwk,
        format: "jwk",
    }),
    transformJwkToKeyObjectAsync: async (jwk) => createPublicKey({
        key: jwk,
        format: "jwk",
    }),
    parseB64UrlString: (b64) => Buffer.from(b64, "base64").toString("utf8"),
    verifySignatureSync: ({ alg, keyObject, jwsSigningInput, signature }) => alg !== "EdDSA"
        ? // eslint-disable-next-line security/detect-object-injection
            createVerify(JwtSignatureAlgorithmHashNames[alg])
                .update(jwsSigningInput)
                .verify({
                key: keyObject,
                dsaEncoding: "ieee-p1363", // Signature format r || s (not used for RSA)
            }, signature, "base64")
        : verify(null, Buffer.from(jwsSigningInput), keyObject, Buffer.from(signature, "base64")),
    verifySignatureAsync: async (args) => nodeWebCompat.verifySignatureSync(args),
    defaultFetchTimeouts: {
        socketIdle: 1500,
        response: 3000,
    },
    setTimeoutUnref: (...args) => setTimeout(...args).unref(),
    transformPemToJwk: async (pem) => {
        return createPublicKey({
            key: Buffer.from(pem),
            format: "pem",
        }).export({
            format: "jwk",
        });
    },
};
