// aws-jwt-verify/dist/cjs/alb-verifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlbJwtVerifier = void 0;
exports.validateAlbJwtFields = validateAlbJwtFields;
exports.validateAndParseAlbArns = validateAndParseAlbArns;
const alb_cache_js_1 = require("./alb-cache.js");
const assert_js_1 = require("./assert.js");
const error_js_1 = require("./error.js");
const jwt_verifier_js_1 = require("./jwt-verifier.js");
const ALB_ARN_REGEX = /^arn:(?:aws|aws-cn|aws-us-gov):elasticloadbalancing:([a-z]{2}-(?:gov-)?[a-z]+-\d{1}):.+$/;
/**
 * Class representing a verifier for JWTs signed by AWS ALB
 */
class AlbJwtVerifier extends jwt_verifier_js_1.JwtVerifierBase {
    constructor(props, jwksCache) {
        const transformPropertiesToIssuerConfig = (props) => {
            const albArns = validateAndParseAlbArns(props.albArn);
            return {
                jwksUri: props.jwksUri ?? getDefaultJwksUri(albArns),
                ...props,
                audience: null,
            };
        };
        const issuerConfig = Array.isArray(props)
            ? props.map(transformPropertiesToIssuerConfig)
            : transformPropertiesToIssuerConfig(props);
        super(issuerConfig, jwksCache);
    }
    static create(verifyProperties, additionalProperties) {
        return new this(verifyProperties, additionalProperties?.jwksCache ?? new alb_cache_js_1.AlbJwksCache());
    }
    /**
     * Verify (synchronously) a JWT that is signed by AWS Application Load Balancer.
     *
     * @param jwt The JWT, as string
     * @param props Verification properties
     * @returns The payload of the JWT––if the JWT is valid, otherwise an error is thrown
     */
    verifySync(...[jwt, properties]) {
        const { decomposedJwt, jwksUri, verifyProperties } = this.getVerifyParameters(jwt, properties);
        try {
            this.verifyDecomposedJwtSync(decomposedJwt, jwksUri, verifyProperties);
            validateAlbJwtFields(decomposedJwt.header, verifyProperties);
        }
        catch (err) {
            if (verifyProperties.includeRawJwtInErrors &&
                err instanceof error_js_1.JwtInvalidClaimError) {
                throw err.withRawJwt(decomposedJwt);
            }
            throw err;
        }
        return decomposedJwt.payload;
    }
    /**
     * Verify (asynchronously) a JWT that is signed by AWS Application Load Balancer.
     * This call is asynchronous, and the JWKS will be fetched from the JWKS uri,
     * in case it is not yet available in the cache.
     *
     * @param jwt The JWT, as string
     * @param props Verification properties
     * @returns Promise that resolves to the payload of the JWT––if the JWT is valid, otherwise the promise rejects
     */
    async verify(...[jwt, properties]) {
        const { decomposedJwt, jwksUri, verifyProperties } = this.getVerifyParameters(jwt, properties);
        try {
            await this.verifyDecomposedJwt(decomposedJwt, jwksUri, verifyProperties);
            validateAlbJwtFields(decomposedJwt.header, verifyProperties);
        }
        catch (err) {
            if (verifyProperties.includeRawJwtInErrors &&
                err instanceof error_js_1.JwtInvalidClaimError) {
                throw err.withRawJwt(decomposedJwt);
            }
            throw err;
        }
        return decomposedJwt.payload;
    }
}
exports.AlbJwtVerifier = AlbJwtVerifier;
function validateAlbJwtFields(header, options) {
    // Check ALB ARN (signer)
    (0, assert_js_1.assertStringArrayContainsString)("ALB ARN", header.signer, options.albArn, error_js_1.AlbJwtInvalidSignerError);
    // Check clientId
    if (options.clientId !== null) {
        if (options.clientId === undefined) {
            throw new error_js_1.ParameterValidationError("clientId must be provided or set to null explicitly");
        }
        (0, assert_js_1.assertStringArrayContainsString)("Client ID", header.client, options.clientId, error_js_1.AlbJwtInvalidClientIdError);
    }
}
function validateAndParseAlbArns(albArn) {
    if (Array.isArray(albArn)) {
        return albArn.map(parseAlbArn);
    }
    else {
        return [parseAlbArn(albArn)];
    }
}
function parseAlbArn(albArn) {
    const match = ALB_ARN_REGEX.exec(albArn);
    if (!match) {
        throw new error_js_1.ParameterValidationError(`Invalid load balancer ARN: ${albArn}`);
    }
    return {
        region: match[1],
    };
}
function getDefaultJwksUri(albArns) {
    const regions = albArns.map((arn) => arn.region);
    const uniqueRegions = Array.from(new Set(regions));
    if (uniqueRegions.length > 1) {
        throw new error_js_1.ParameterValidationError("Using ALBs from different regions is not supported for the same issuer");
    }
    return `https://public-keys.auth.elb.${uniqueRegions[0]}.amazonaws.com`;
}
