// aws-jwt-verify/dist/cjs/cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleLruCache = void 0;
class SimpleLruCache {
    constructor(capacity) {
        this.capacity = capacity;
        if (capacity < 1) {
            throw new Error(`capacity must be greater than 0, but got ${capacity}`);
        }
        this.index = new Map();
    }
    get size() {
        return this.index.size;
    }
    get(key) {
        const value = this.index.get(key);
        if (value) {
            this.moveFirst(key, value);
            return value;
        }
        else {
            return undefined;
        }
    }
    set(key, value) {
        if (this.size >= this.capacity) {
            this.removeLast();
        }
        this.moveFirst(key, value);
        return this;
    }
    moveFirst(key, value) {
        this.index.delete(key);
        this.index.set(key, value);
    }
    removeLast() {
        const last = this.index.keys().next().value;
        if (last) {
            this.index.delete(last);
        }
    }
    /**
     *
     * @returns array ordered from the least recent to the most recent
     */
    toArray() {
        return Array.from(this.index);
    }
}
exports.SimpleLruCache = SimpleLruCache;
